### PR TITLE
Handle BadHeaderError in AdminEmailHandler

### PR DESCRIPTION
**Pull request by sravfeyn**
http://manage.dimagi.com/default.asp?178559#edit_0_178559

### CODE REVIEW

File: corehq/util/log.py [hunk 1]
```diff
@@ -131,6 +131,18 @@ def get_code(self, extracted_tb):
...
+    def _clean_subject(cls, subject):
+        # Django raises BadHeaderError if subject contains following bad_strings
+        # to guard against Header Inejction.
+        # see https://docs.djangoproject.com/en/1.8/topics/email/#preventing-header-injection
+        # bad-strings list from http://nyphp.org/phundamentals/8_Preventing-Email-Header-Injection
+        bad_strings = ["\r", "\n", "%0a", "%0d", "Content-Type:", "bcc:", "to:", "cc:"]
+        replacement = "-"
+        for i in bad_strings:
+            subject.replace(i, replacement)
+        retun subject
```
Comment (dimagimon) [inline corehq/util/log.py]: E901 SyntaxError: invalid syntax

**Comment by sravfeyn** [comment]
cc @esoergel

File: corehq/util/log.py (hunk 1 above)
Comment (sravfeyn) [inline corehq/util/log.py]: Oops

File: corehq/util/log.py [hunk 2]
```diff
@@ -131,6 +131,18 @@ def get_code(self, extracted_tb):
         )
         )
 
+    @classmethod
+    def _clean_subject(cls, subject):
+        # Django raises BadHeaderError if subject contains following bad_strings
+        # to guard against Header Inejction.
+        # see https://docs.djangoproject.com/en/1.8/topics/email/#preventing-header-injection
+        # bad-strings list from http://nyphp.org/phundamentals/8_Preventing-Email-Header-Injection
+        bad_strings = ["\r", "\n", "%0a", "%0d", "Content-Type:", "bcc:", "to:", "cc:"]
```
Comment (sravfeyn) [inline corehq/util/log.py]: Not that we would add these strings in the code, but I wanted to put an exhaustive list possible.

File: corehq/util/log.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def emit(self, record):
         context = self.get_context(record)
 
         message = "\n\n".join(filter(None, [
-            context['message'],
+            self._clean_subject(context['message']),
```
Comment (esoergel) [inline corehq/util/log.py]: I think you're cleaning the wrong thing here - this is the message, which goes in the body of the email.  On line 105 here, `mail.mail_admins` is passed `context['subject']` as the first arg - I believe that's the problem area.

Comment (sravfeyn) [inline corehq/util/log.py]: Ugh, definitely reminded myself to be careful there and forgot. Thanks. Fixing.

**Comment by esoergel** [comment]
I should have mentioned this in the ticket, but these changes are hard to test - to get the emails to show up in your shell, you need to:

Comment out this filter:
https://github.com/dimagi/commcare-hq/blob/a741d1edfd228c16ccdd6b44c966956653890944/settings.py#L830
and make this line `if False:`
https://github.com/dimagi/commcare-hq/blob/a741d1edfd228c16ccdd6b44c966956653890944/settings.py#L952
and this line should be in your localsettings if it's not already:
```python
EMAIL_BACKEND='django.core.mail.backends.console.EmailBackend'
```

**Comment by czue** [comment]
(you shouldn't need to comment out that line, just add `FIX_LOGGER_ERROR_OBFUSCATION = False` to localsettings)

File: corehq/util/log.py [hunk 4]
```diff
@@ -131,6 +131,18 @@ def get_code(self, extracted_tb):
...
         )
 
+    @classmethod
+    def _clean_subject(cls, subject):
+        # Django raises BadHeaderError if subject contains following bad_strings
+        # to guard against Header Inejction.
+        # see https://docs.djangoproject.com/en/1.8/topics/email/#preventing-header-injection
+        # bad-strings list from http://nyphp.org/phundamentals/8_Preventing-Email-Header-Injection
+        bad_strings = ["\r", "\n", "%0a", "%0d", "Content-Type:", "bcc:", "to:", "cc:"]
+        replacement = "-"
```
Comment (czue) [inline corehq/util/log.py]: subjective but i'd use a space character - especially if newlines are the most common thing

**Comment by czue** [comment]
(no blockers from me)

**Pull request merged**